src/logits_processor.cpp

```diff
@@ -0,0 +1,240 @@
+      &tokenize_data_,                     // user_data
+  };
+
+  char error_buf[128];
+  llg_tokenizer_ = std::unique_ptr<LlgTokenizer, LlgTokenizerDeleter>(llg_new_tokenizer(&tokenizer_init, error_buf, sizeof(error_buf)));
+  if (!llg_tokenizer_) {
+    throw std::runtime_error("Error creating llg_tokenizer: " + std::string(error_buf));
+  }
+
+  llg_constraints_.resize(batch_size_);
+  for (int i = 0; i < batch_size_; i++) {
+    LlgConstraintInit constraint_init;
+    llg_constraint_init_set_defaults(&constraint_init, llg_tokenizer_.get());
+    LlgConstraint* constraint_ptr;
+    if (guidance_type_ == "json_schema") {
+      constraint_ptr = llg_new_constraint_json(&constraint_init, guidance_data_.data());
+    } else if (guidance_type_ == "regex") {
+      constraint_ptr = llg_new_constraint_regex(&constraint_init, guidance_data_.data());
+    } else {
+      throw std::runtime_error("Unsupported guidance type: " + std::string(guidance_type_) + " (only json_schema and regex are supported)");
+    }
+    if (llg_get_error(constraint_ptr) != nullptr) {
+      std::string error_message = llg_get_error(constraint_ptr);
+      llg_free_constraint(constraint_ptr);
+      throw std::runtime_error("Error creating grammar: " + error_message);
+    }
+    llg_constraints_[i] = std::unique_ptr<LlgConstraint, LlgConstraintDeleter>(constraint_ptr);
+  }
+
+  // Compute the mask asynchronously to avoid blocking the model inference on device
+  mask_future_ = std::async(std::launch::async, [&]() {
+    return ComputeMask();
+  });
+
+#if USE_CUDA
+  if (state.params_->device_type == DeviceType::CUDA) {
+    cuda_logits_mask_ptr_ = state.params_->p_device->Allocate<uint32_t>(batch_size_ * vocab_size_ / 32);
+  }
+  cuda_stream_ = state.params_->cuda_stream;
+#endif
+}
+
+std::vector<std::vector<uint32_t>> GuidanceLogitsProcessor::ComputeMask() {
+  std::vector<std::vector<uint32_t>> masks;
+  for (int batch_idx = 0; batch_idx < batch_size_; batch_idx++) {  // renamed 'i' to 'batch_idx'
+    LlgMaskResult mask_result;
+    auto error = llg_compute_mask(llg_constraints_[batch_idx].get(), &mask_result);
+    if (error != 0) {
+      std::string error_message = llg_get_error(llg_constraints_[batch_idx].get());
+      throw std::runtime_error("Error computing mask: " + error_message);
+    }
+
+    std::vector<uint32_t> mask;
+    if (mask_result.is_stop) {
+      mask = std::vector<uint32_t>((vocab_size_ - 1) / 32 + 1, 0);
+      uint32_t eos_mask32 = 1 << (eos_token_ % 32);
+      mask[eos_token_ / 32] = eos_mask32;
+    } else {
+      mask.reserve((vocab_size_ - 1) / 32 + 1);
+      for (int i = 0; i < (vocab_size_ - 1) / 32 + 1; i++) {
+        mask.push_back(mask_result.sample_mask[i]);
+      }
+    }
+    masks.push_back(mask);
+  }
+  return masks;
+}
+
+void GuidanceLogitsProcessor::CommitTokens(std::span<int32_t> tokens) {
+  for (int i = 0; i < batch_size_; i++) {
+    LlgCommitResult commit_result;
```

Comment:
What is commit_result used for, if llg_commit_token returns an error?

Reply:
commit_result is useless for now. Jump forward feature will return multiple tokens that is fixed in next steps, so we can call append_tokens(commit_result.tokens) to accelerate constrianed decoding. I will add a new PR to support this.